fineract-provider/src/main/java/org/apache/fineract/infrastructure/security/filter/FineractInstanceModeApiFilter.java

```diff
@@ -0,0 +1,86 @@
+ * under the License.
+ */
+package org.apache.fineract.infrastructure.security.filter;
+
+import java.io.IOException;
+import javax.servlet.FilterChain;
+import javax.servlet.ServletException;
+import javax.servlet.ServletRequest;
+import javax.servlet.ServletResponse;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import javax.ws.rs.HttpMethod;
+import javax.ws.rs.ext.Provider;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.commons.lang3.time.StopWatch;
+import org.apache.fineract.infrastructure.core.config.FineractProperties;
+import org.apache.fineract.infrastructure.core.data.ApiGlobalErrorResponse;
+import org.apache.fineract.infrastructure.core.serialization.ToApiJsonSerializer;
+import org.apache.fineract.infrastructure.security.data.PlatformRequestLog;
+import org.apache.http.HttpStatus;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
+import org.springframework.web.filter.GenericFilterBean;
+
+@Provider
+@Component
+@Slf4j
+public class FineractInstanceModeApiFilter extends GenericFilterBean {
+
+    @Autowired
+    private FineractProperties fineractProperties;
+
+    @Autowired
+    private ToApiJsonSerializer<PlatformRequestLog> toApiJsonSerializer;
+
+    @Override
+    public void doFilter(final ServletRequest req, final ServletResponse res, final FilterChain chain)
+            throws IOException, ServletException {
+
+        final HttpServletRequest request = (HttpServletRequest) req;
+        final HttpServletResponse response = (HttpServletResponse) res;
+
+        log.debug("{} {} {} {}", request.getMethod(), fineractProperties.getMode().isReadEnabled(),
+                fineractProperties.getMode().isWriteEnabled(), fineractProperties.getMode().isBatchEnabled());
+
+        final StopWatch task = new StopWatch();
+        task.start();
+
+        if (fineractProperties.getMode().isReadInstance() && isReadMethod(request)) {
+            chain.doFilter(request, response);
+        } else if (fineractProperties.getMode().isWriteInstance() && isWriteMethod(request)) {
+            chain.doFilter(request, response);
+        } else {
+            task.stop();
+            final PlatformRequestLog platformRequestLog = PlatformRequestLog.from(task, request);
+            log.debug("{}", this.toApiJsonSerializer.serialize(platformRequestLog));
+            response.setStatus(HttpStatus.SC_METHOD_NOT_ALLOWED);
+            ApiGlobalErrorResponse errorResponse = ApiGlobalErrorResponse.invalidInstanceTypeMethod(request.getMethod());
+            response.getWriter().write(errorResponse.toJson());
+        }
+    }
+
+    private boolean isReadMethod(HttpServletRequest request) {
+        return HttpMethod.GET.equals(request.getMethod());
+    }
+
+    private boolean isWriteMethod(HttpServletRequest request) {
+        return !HttpMethod.GET.equals(request.getMethod());
```

Comment:
Maybe negating the isReadMethod is more readable, like `!isReadMethod()`

Reply:
Done